src/NitroSharp/InputContext.cs

```diff
@@ -65,38 +72,52 @@ public void Update(SystemVariableLookup systemVariables)
             MousePosition = input.Snapshot.MousePosition;
             WheelDelta = input.WheelDelta;
 
-            PollVkey(input, VirtualKey.Enter);
-            PollVkey(input, VirtualKey.Advance);
-            PollVkey(input, VirtualKey.Back);
-            PollVkey(input, VirtualKey.Left);
-            PollVkey(input, VirtualKey.Up);
-            PollVkey(input, VirtualKey.Right);
-            PollVkey(input, VirtualKey.Down);
-            PollVkey(input, VirtualKey.Skip);
+            foreach (VirtualKey key in Enum.GetValues<VirtualKey>())
+            {
+                PollVkey(input, key);
+            }
 
             Gamepad gamepad = input.Gamepad;
             SystemVariableLookup sys = systemVariables;
-            set(ref sys.RightButtonDown, VKeyState(VirtualKey.Back));
-            pollController(SDL_GameControllerButton.Start, ref sys.X360StartButtonDown);
-            set(ref sys.X360AButtonDown, VKeyState(VirtualKey.Advance));
-            set(ref sys.X360BButtonDown, VKeyState(VirtualKey.Back));
-            pollController(SDL_GameControllerButton.Y, ref sys.X360YButtonDown);
-            set(ref sys.X360LeftButtonDown, VKeyState(VirtualKey.Left));
-            set(ref sys.X360UpButtonDown, VKeyState(VirtualKey.Down));
-            set(ref sys.X360RightButtonDown, VKeyState(VirtualKey.Right));
-            set(ref sys.X360DownButtonDown, VKeyState(VirtualKey.Down));
-            pollController(SDL_GameControllerButton.LeftShoulder, ref sys.X360LbButtonDown);
-            pollController(SDL_GameControllerButton.RightShoulder, ref sys.X360RbButtonDown);
+
+            setDown(ref sys.X360AButtonDown, VirtualKey.Advance);
+            setDown(ref sys.X360BButtonDown, VirtualKey.Back);
+            setDownUp(ref sys.X360YButtonDown, VirtualKey.Y);
+
+            setDownUp(ref sys.X360UpButtonDown, VirtualKey.Up);
+            setDownUp(ref sys.X360DownButtonDown, VirtualKey.Down);
+            setDownUp(ref sys.X360LeftButtonDown, VirtualKey.Left);
+            setDownUp(ref sys.X360RightButtonDown, VirtualKey.Right);
+
+            setDownUp(ref sys.X360StartButtonDown, VirtualKey.Start);
+            setDownUp(ref sys.X360LbButtonDown, VirtualKey.BumperLeft);
+            setDownUp(ref sys.X360RbButtonDown, VirtualKey.BumperRight);
 
             static void set(ref ConstantValue target, bool value)
             {
                 target = ConstantValue.Boolean(value);
             }
 
-            void pollController(SDL_GameControllerButton button, ref ConstantValue val)
+            void setDown(ref ConstantValue target, VirtualKey key)
             {
-                bool down = gamepad.ButtonState(button);
-                set(ref val, down);
+                if (VKeyDown(key))
+                {
+                    set(ref target, true);
+                }
+            }
+
+            void setUp(ref ConstantValue target, VirtualKey key)
```

Comment:
Why is it necessary to have seperate ``setUp`` and ``setDown`` methods? Are you implying that the variable corresponding to a given key is supposed to remain ``null`` until it's first pressed?

Reply:
For variables corresponding to some specific keys, N2 only ever sets them to `true` (they should be manually reset to `false` in the script), that's what `setDown` is for.